rclpy/src/rclpy_common/src/common.c

```diff
@@ -34,63 +34,56 @@ rclpy_convert_to_py_qos_policy(void * profile)
     return NULL;
   }
 
+  // start qos setting
   rmw_qos_profile_t * qos_profile = (rmw_qos_profile_t *)profile;
-
-  PyObject * pyqos_depth = PyLong_FromSize_t(qos_profile->depth);
+  PyObject * pyqos_depth = NULL;
+  PyObject * pyqos_history = NULL;
+  PyObject * pyqos_reliability = NULL;
+  PyObject * pyqos_durability = NULL;
+  PyObject * pyqos_avoid_ros_namespace_conventions = NULL;
```

Comment:
You could declare a `rclpy_row_profile_t` which would be a 1:1 mapping of `rmw_qos_profile_t` with PyObject.
This would allow you to e.g. pass it to a cleaning function and avoid the goto.

Reply:
I'm not positive I understand - you mean something like

```
struct rclpy_qos_profile_t {
  PyObject * depth;
  PyObject * history;
  // etc ...
};

void clean_rclpy_qos_profile(rclpy_qos_profile_t * p) {
  Py_XDECREF(p->depth);
  Py_XDECREF(p->history);
  // etc ...
}

PyObject * rclpy_convert_to_py_qos_policy(void * profile) {
  rclpy_qos_profile_t pyprof;
  // etc ...

  pyprof->depth = PyLong_FromSize_t(qos_profile->depth);
  if (!pyprof->depth) {
    clean_rclpy_qos_profile(pyprof);
    return NULL;
  }

  //etc ...
}

```

Reply:
Yeah, having a separate clean-up function would spare you the need to use goto.
It's also easier to test, easier to extend, easier to split into multiple functions, etc.